### PR TITLE
Introduce a concept of a linking environment for wasmate.

### PR DESCRIPTION
**Pull request by sunfishcode**
This is inspired by https://github.com/WebAssembly/spec/pull/196. It significantly revamps how imports are handled in wasmate.py.

.s files no longer declare their imports with .imports and .import . They just use symbols without defining them, as is conventional in .s files.

wasmate.py now has collections of imports, which is a stand-in for what it would get from linking with a  real library. It uses these to resolve the symbols in the .s file into proper imports, complete with a module name.

This would be accompanied by an LLVM change to delete the EmitEndOfAsmFile function in WebAssemblyAsmPrinter.cpp, which is the code that emits .imports and .import directives.

### CODE REVIEW

**Comment by kripken** [comment]
Just using symbols without defining them is ok in various assemblies because they are not strongly typed, isn't it? By which I mean, they tolerate calling a function with a wrong argument type of # of parameters. But in wasm, we are strict there, so having an import define feels natural to me.

**Comment by sunfishcode** [comment]
Wasm engines are going to have to do signature checking for every static callsite in any case; this seems unavoidable. The question here seems to be whether we want an additional round of signature checking at link time.

An additional round would have the advantage of diagnosing most type mismatches earlier in a typical workflow. It would have the disadvantage of adding more for data for a linker to read and work for a linker to do.

My expectation is that signature mismatches at link time are very infrequent in practice. And, my expectation is that the linking stage is very likely to be a compile-time bottleneck.

Signature mismatches will happen occasionally, but they will still be diagnosed in an explicit way even if we don't do it in the link step; they'll just be reported a little later. This seems to be a reasonable tradeoff to potentially speed up and simplify the link step.

File: prototype-wasmate/wasmate.py [hunk 1]
```diff
@@ -5,6 +5,51 @@
...
+    'printf': ('printf', 'misctest', 'f32', 'f32'),
+    'printi': ('printi', 'misctest', 'i32', 'i32'),
+    'printv': ('printv', 'misctest', '', ''),
+    'return_something': ('return_something', 'misctest', '', 'i32'),
+    'something': ('something', 'misctest', '', ''),
+    'split_arg': ('split_arg', 'misctest', 'i64 i64', ''),
+    'void_nullary': ('void_nullary', 'misctest', '', ''),
+    '_ZN5AppleC1Ev': ('_ZN5AppleC1Ev', 'misctest', 'i32', 'i32'),
+    '_Znwm': ('_Znwm', 'misctest', 'i32', 'i32'),
+}
```
Comment (jfbastien) [inline prototype-wasmate/wasmate.py]: Are these basically all the unresolved symbols that we have in LLVM tests? The syscall-y C library ones seem easy enough to keep here, but the others don't really seem to fit. I'm running the torture test output through these right now and I'm hoping we can get something a bit more generic: abort and puts make sense, but I'm not sure about the others.

WDYT?

**Comment by jfbastien** [comment]
I like the idea, but I think it should be provided in external files instead of hard-coded.

**Comment by sunfishcode** [comment]
Yeah, the misctest_environment set is just the set of things that happen to be used in the current set of tests. It's not meant to be anything that would actually be implemented by a real host.

I'm open to moving these data structures to a separate file. They're intended to simulate the information that would come from a real library, so that would be a step closer. That said, would you mind if we defer that for the moment and file an issue for it?

The big question at this moment is whether we want imports to work like this, in general, so I'd like to finish that discussion before taking the next step of inventing/discovering a file format to describe module interfaces.

**Comment by jfbastien** [comment]
As discussed on IRC, this boils down to:
* Should this be in the C++ header (using `__attribute__` or some asm thing).
* Should this be in a `.def` file, with names and types?
In either case it can be a standard file, and devs can have their own.

@sunfishcode was suggesting that this could be related to ES6 module imports. Maybe? I don't want to tie this to web stuff necessarily, but we should do *something* and then revisit when the dust settles.

I'm not opposed to a quick hack for now.

**Comment by jfbastien** [comment]
As discussed on IRC, it's probably fine to move forward with this hack and revisit later.

**Comment by jfbastien** [comment]
Also, I think the long-term solutions I'd lean towards is in header files:
```
void foo(void) asm("my_asm_name") __attribute__((wasm_export("my_other_name")));
```

**Comment by sunfishcode** [comment]
@jfbastien Indeed; that's the syntax we discussed on IRC. We'll need to design it carefully. Until then, this PR moves us forward with a system that is easy to work with, and that we may want to keep even if we do have wasm_import/wasm_export.

**Pull request merged**